File: P2/clusters/Hierarchical.py
# coding = utf8

import numpy as np

class ClusterNode:
    def __init__(self, center, left=None, right=None, distance=-1, id=None, count=1):
        self.center = center
        self.left = left
        self.right = right
        self.distance= distance
        self.id = id
        self.count = count

class HierarchicalClustering:
    def __init__(self, n_clusters):
        self.__n_clusters = n_clusters
        self.__dataset = None
    
    def __distance(self, x, y):
        return np.sqrt(np.sum((x - y) ** 2))

    def __traverse(self, node, label):
        if node.left is None and node.right is None:
            self.label_pred[node.id] = label
        if node.left:
            self.__traverse(node.left, label)
        if node.right:
            self.__traverse(node.right, label)

    def __set_labels(self, tree):
        for i in range(self.__n_clusters):
            self.__traverse(tree[i], i)

    def __cluster(self):
        m, n = np.shape(self.__dataset)
        distance = dict()
        nodes = [ClusterNode(self.__dataset[i], id=i) for i in range(m)]
        cluster_id = -1

        # cluster process
        while len(nodes) > self.__n_clusters:
            min_dis = float("inf")
            length = len(nodes)
            closest = None
            
            # 寻找距离最近的两个聚类
            for i in range(length - 1):
                for j in range(i + 1, length):
                    key = (nodes[i].id, nodes[j].id)
                    if key not in distance:
                        distance[key] = self.__distance(nodes[i].center, nodes[j].center)
                    if distance[key] < min_dis:
                        min_dis = distance[key]
                        closest = (i, j)

            # 合并两个聚类
            node1, node2 = nodes[closest[0]], nodes[closest[1]]
            new_point = np.array([(
                node1.center[i] * node1.count + node2.center[i] * node2.count) / (node1.count + node2.count) for i in range(n)])
            new_node = ClusterNode(new_point, 
                                   left=node1, 
                                   right=node2,
                                   distance=min_dis,
                                   id=cluster_id,
                                   count=node1.count + node2.count)
            cluster_id -= 1
            del nodes[closest[1]], nodes[closest[0]]
            nodes.append(new_node)
            print('done for %d' % cluster_id)

        self.label_pred = [-1] * m
        self.__set_labels(nodes)

    def fit(self, dataset):
        self.__dataset = dataset
        self.__cluster()
        return

File: P2/clusters/Hierarchical.cpp
#include<iostream>
#include<vector>
#include<map>
#include<list>
#include<math.h>
#include<climits>

using namespace std;

class ClusterNode {
public:
    vector<double> center;
    ClusterNode* left;
    ClusterNode* right;
    double distance;
    int id;
    int count;
    ClusterNode(vector<double> center, int id = INT_MIN, ClusterNode* left = nullptr, 
        ClusterNode* right = nullptr, double distance = -1, int count = 1) {
        this -> center = center;
        this -> left = left;
        this -> right = right;
        this -> distance = distance;
        this -> id = id;
        this -> count = count;
    }
    ~ClusterNode() {}
};

class HierarchicalClustering {
private:
    int n_clusters;
    vector<vector<double>> dataset;
    int m, n;
public:
    vector<int> label_pred;

private:
    double distance(vector<double> x, vector<double> y) {
        double sum = 0;
        for (int i = 0; i < this -> n; i++)
            sum += (x[i] - y[i]) * (x[i] - y[i]);
        return sqrt(sum);
    }

    void traverse(ClusterNode node, int label) {
        if (node.left == nullptr && node.right == nullptr)
            this -> label_pred[node.id] = label;
        if (node.left) traverse(*node.right, label);
        if (node.right) traverse(*node.right, label);
    }

    void set_labels(list<ClusterNode> tree) {
        int label = 0;
        for (auto p = tree.begin(); p != tree.end(); p++)
            traverse(*p, label);
    }

    void cluster() {
        m = dataset.size();
        n = dataset[0].size();
        map<pair<int, int>, double> distance;
        list<ClusterNode> nodes;
        for (int i = 0; i < m; i++) {
            nodes.push_back(ClusterNode(dataset[i], i));
        }
        int cluster_id = -1;

        // while (nodes.size() > n_clusters) {
        //     double min_dis = INFINITY;
        //     int length = nodes.size();
        //     pair<int, int> clostest;
        //     for (int i = 0; i < length - 1; i++)
        //         for (int j = i + 1; j < length; j++) {
        //             pair<int ,int> key(nodes[i].id, nodes[j].id);
        //             auto p = distance.find(pair<int, int>(, j));
        //             double dis = 0;
        //             if (p != distance.end()) dis = p->second;
        //             else distance.insert(make_pair(pair<int, int>(i, j)))
        //         }
            
        // }
    }
public:
    HierarchicalClustering(int n_clusters) {
        this -> n_clusters =  n_clusters;
    }
    ~HierarchicalClustering() {}

    void fit(vector<vector<double>> dataset) {
        this->dataset = dataset;
        cluster();
    }
};
